File: android/src/main/java/com/scandit/datacapture/flutter/barcode/selection/BarcodeSelectionMethodHandler.java
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2024- Scandit AG. All rights reserved.
 */
package com.scandit.datacapture.flutter.barcode.selection;

import androidx.annotation.NonNull;

import com.scandit.datacapture.flutter.core.utils.FlutterResult;
import com.scandit.datacapture.flutter.core.utils.ResultUtils;
import com.scandit.datacapture.frameworks.barcode.selection.BarcodeSelectionModule;
import com.scandit.datacapture.frameworks.core.FrameworkModule;
import com.scandit.datacapture.frameworks.core.errors.FrameDataNullError;
import com.scandit.datacapture.frameworks.core.locator.ServiceLocator;
import com.scandit.datacapture.frameworks.core.utils.DefaultLastFrameData;
import com.scandit.datacapture.frameworks.core.utils.LastFrameData;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

import org.json.JSONObject;

public class BarcodeSelectionMethodHandler implements MethodChannel.MethodCallHandler {

    public static final String EVENT_CHANNEL_NAME = "com.scandit.datacapture.barcode.selection/event_channel";
    public static final String METHOD_CHANNEL_NAME = "com.scandit.datacapture.barcode.selection/method_channel";

    private final ServiceLocator<FrameworkModule> serviceLocator;
    private final LastFrameData lastFrameData;

    public BarcodeSelectionMethodHandler(ServiceLocator<FrameworkModule> serviceLocator) {
        this(serviceLocator, DefaultLastFrameData.getInstance());
    }

    public BarcodeSelectionMethodHandler(ServiceLocator<FrameworkModule> serviceLocator, LastFrameData lastFrameData) {
        this.serviceLocator = serviceLocator;
        this.lastFrameData = lastFrameData;
    }

    @Override
    public void onMethodCall(MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "getBarcodeSelectionDefaults":
                result.success(new JSONObject(getSharedModule().getDefaults()).toString());
                break;

            case "getBarcodeSelectionSessionCount":
                assert call.arguments() != null;
                result.success(getSharedModule().getBarcodeCount(call.arguments()));
                break;

            case "resetBarcodeSelectionSession":
                getSharedModule().resetLatestSession(call.arguments());
                result.success(null);
                break;

            case "addBarcodeSelectionListener":
                getSharedModule().addListener();
                result.success(null);
                break;

            case "removeBarcodeSelectionListener":
                getSharedModule().removeListener();
                result.success(null);
                break;

            case "resetMode":
                getSharedModule().resetSelection();
                result.success(null);
                break;

            case "unfreezeCamera":
                getSharedModule().unfreezeCamera();
                result.success(null);
                break;

            case "finishDidUpdateSelection":
                getSharedModule().finishDidSelect(Boolean.TRUE.equals(call.arguments()));
                result.success(null);
                break;

            case "finishDidUpdateSession":
                getSharedModule().finishDidUpdateSession(
                        Boolean.TRUE.equals(call.arguments())
                );
                result.success(null);
                break;

            case "getLastFrameData":
                lastFrameData.getLastFrameDataBytes(bytes -> {
                    if (bytes == null) {
                        ResultUtils.rejectKotlinError(result, new FrameDataNullError());
                        return null;
                    }
                    result.success(result);
                    return null;
                });
                break;

            case "setModeEnabledState":
                getSharedModule().setModeEnabled(Boolean.TRUE.equals(call.arguments()));
                break;

            case "updateBarcodeSelectionMode":
                assert call.arguments() != null;
                getSharedModule().updateModeFromJson(
                        call.arguments(),
                        new FlutterResult(result)
                );
                break;

            case "applyBarcodeSelectionModeSettings":
                assert call.arguments() != null;
                getSharedModule().applyModeSettings(
                        call.arguments(),
                        new FlutterResult(result)
                );
                break;

            case "updateBarcodeSelectionBasicOverlay":
                assert call.arguments() != null;
                getSharedModule().updateBasicOverlay(
                        call.arguments(),
                        new FlutterResult(result)
                );
                break;
            default:
                throw new IllegalStateException("Unexpected value: " + call.method);
        }
    }

    private volatile BarcodeSelectionModule sharedModuleInstance;

    private BarcodeSelectionModule getSharedModule() {
        if (sharedModuleInstance == null) {
            synchronized (this) {
                if (sharedModuleInstance == null) {
                    sharedModuleInstance = (BarcodeSelectionModule)this.serviceLocator.resolve(BarcodeSelectionModule.class.getName());
                }
            }
        }
        return sharedModuleInstance;
    }
}

File: android/src/main/kotlin/com/scandit/datacapture/flutter/barcode/selection/BarcodeSelectionMethodHandler.kt
/*
 * This file is part of the Scandit Data Capture SDK
 *
 * Copyright (C) 2021- Scandit AG. All rights reserved.
 */

package com.scandit.datacapture.flutter.barcode.selection

import com.scandit.datacapture.flutter.core.utils.FlutterResult
import com.scandit.datacapture.flutter.core.utils.rejectKotlinError
import com.scandit.datacapture.frameworks.barcode.selection.BarcodeSelectionModule
import com.scandit.datacapture.frameworks.core.errors.FrameDataNullError
import com.scandit.datacapture.frameworks.core.utils.DefaultLastFrameData
import com.scandit.datacapture.frameworks.core.utils.LastFrameData
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

class BarcodeSelectionMethodHandler(
    private val barcodeSelectionModule: BarcodeSelectionModule,
    private val lastFrameData: LastFrameData = DefaultLastFrameData.getInstance()
) : MethodChannel.MethodCallHandler {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            METHOD_GET_BARCODE_SELECTION_DEFAULTS ->
                result.success(JSONObject(barcodeSelectionModule.getDefaults()).toString())

            METHOD_GET_BARCODE_COUNT ->
                result.success(
                    barcodeSelectionModule.getBarcodeCount(
                        call.arguments as String
                    )
                )

            METHOD_RESET_BARCODE_SELECTION_SESSION -> {
                barcodeSelectionModule.resetLatestSession(call.arguments as? Long)
                result.success(null)
            }

            METHOD_ADD_LISTENER -> {
                barcodeSelectionModule.addListener()
                result.success(null)
            }

            METHOD_REMOVE_LISTENER -> {
                barcodeSelectionModule.removeListener()
                result.success(null)
            }

            METHOD_RESET_MODE -> {
                barcodeSelectionModule.resetSelection()
                result.success(null)
            }

            METHOD_UNFREEZE_CAMERA -> {
                barcodeSelectionModule.unfreezeCamera()
                result.success(null)
            }

            METHOD_BARCODE_SELECTION_DID_SELECT -> {
                val enabled = call.arguments as? Boolean ?: false
                barcodeSelectionModule.finishDidSelect(enabled)
                result.success(null)
            }

            METHOD_BARCODE_SELECTION_DID_UPDATE_SESSION -> {
                val enabled = call.arguments as? Boolean ?: false
                barcodeSelectionModule.finishDidUpdateSession(enabled)
                result.success(null)
            }

            METHOD_GET_LAST_FRAME_DATA -> lastFrameData.getLastFrameDataBytes {
                if (it == null) {
                    result.rejectKotlinError(FrameDataNullError())
                    return@getLastFrameDataBytes
                }
                result.success(result)
            }

            METHOD_SET_MODE_ENABLED_STATE -> barcodeSelectionModule.setModeEnabled(
                call.arguments as Boolean
            )

            METHOD_UPDATE_MODE_FROM_JSON -> barcodeSelectionModule.updateModeFromJson(
                call.arguments as String,
                FlutterResult(result)
            )

            METHOD_APPLY_MODE_SETTINGS -> barcodeSelectionModule.applyModeSettings(
                call.arguments as String,
                FlutterResult(result)
            )

            METHOD_UPDATE_BASIC_OVERLAY -> barcodeSelectionModule.updateBasicOverlay(
                call.arguments as String,
                FlutterResult(result)
            )
        }
    }

    companion object {
        private const val METHOD_GET_BARCODE_SELECTION_DEFAULTS = "getBarcodeSelectionDefaults"
        private const val METHOD_GET_BARCODE_COUNT = "getBarcodeSelectionSessionCount"
        private const val METHOD_RESET_BARCODE_SELECTION_SESSION = "resetBarcodeSelectionSession"
        private const val METHOD_ADD_LISTENER = "addBarcodeSelectionListener"
        private const val METHOD_REMOVE_LISTENER = "removeBarcodeSelectionListener"
        private const val METHOD_RESET_MODE = "resetMode"
        private const val METHOD_UNFREEZE_CAMERA = "unfreezeCamera"
        private const val METHOD_BARCODE_SELECTION_DID_SELECT = "finishDidUpdateSelection"
        private const val METHOD_BARCODE_SELECTION_DID_UPDATE_SESSION = "finishDidUpdateSession"
        private const val METHOD_GET_LAST_FRAME_DATA = "getLastFrameData"
        private const val METHOD_SET_MODE_ENABLED_STATE = "setModeEnabledState"
        private const val METHOD_UPDATE_MODE_FROM_JSON = "updateBarcodeSelectionMode"
        private const val METHOD_APPLY_MODE_SETTINGS = "applyBarcodeSelectionModeSettings"
        private const val METHOD_UPDATE_BASIC_OVERLAY = "updateBarcodeSelectionBasicOverlay"

        const val EVENT_CHANNEL_NAME =
            "com.scandit.datacapture.barcode.selection/event_channel"

        const val METHOD_CHANNEL_NAME =
            "com.scandit.datacapture.barcode.selection/method_channel"
    }
}
